Add a new HollowUniqueKeyIndex that supports object longevity. Update HollowHashIndex to support object longevity.

Jorel-Master-Of-Scheduling opened the pull request:
Add an alternative to HollowPrimaryKeyIndex called HollowUniqueKeyIndex. It uses HollowDataAccess instead of HollowObjectTypeReadState. This allows HollowUniqueKeyIndex to opt-in to Object Longevity by using the HollowDataAccess from a Consumer rather than the HollowObjectTypeReadState.

HollowObjectTypeReadState is mutable and gets updated with each delta.  If a PrimaryKeyIndex does NOT listen to updates, then it will become stale over time. The internal index structure will remain unchanged, but the HollowPrimaryKeyValueDeriver will start returning incorrect keys as the ordinals get re-used and re-assigned to new records.

By migrating to HollowDataAccess, users can choose to use the HollowObjectTypeReadState (which stays up to date) or use HollowApi.getDataAccess() which may be object-longevity-aware.

Jorel-Master-Of-Scheduling committed:
Update PrimaryKeyIndex to use HollowDataAccess instead of HollowObjectTypeReadState. This allows PrimaryKeyIndex to opt-in to Object Longevity by using the HollowDataAccess from a Consumer rather than the HollowObjectTypeReadState.

HollowObjectTypeReadState is mutable and gets updated with each delta.  If a PrimaryKeyIndex does NOT listen to updates, then it will become stale over time. The internal index structure will remain unchanged, but the HollowPrimaryKeyValueDeriver will start returning incorrect keys as the ordinals get re-used and re-assigned to new records.

By migrating to HollowDataAccess, users can choose to use the HollowObjectTypeReadState (which stays up to date) or use HollowApi.getDataAccess() which may be object-longevity-aware.

Jorel-Master-Of-Scheduling committed:
Update HollowHashIndex to use HollowDataAccess instead of HollowObjectTypeReadState. This allows HollowHashIndex to opt-in to Object Longevity by using the HollowDataAccess from a Consumer rather than the HollowObjectTypeReadState.

HollowObjectTypeReadState is mutable and gets updated with each delta.  If a HollowHashIndex does NOT listen to updates, then it will become stale over time. Unfortunately, when using the StateEngine, the staleness causes the ordinals to mismatch against the state the index was built against.

By migrating to HollowDataAccess, users can choose to use the HollowObjectTypeReadState (which stays up to date) or use HollowApi.getDataAccess() which may be object-longevity-aware.

Jorel-Master-Of-Scheduling committed:
Fix type data accessor lookups for collection and map.

Jorel-Master-Of-Scheduling committed:
Add addition tests to HollowHashIndexTest.

Jorel-Master-Of-Scheduling committed:
Refactor out most of the common code between HollowPrimaryKeyIndex and HollowUniqueKeyIndex. Add test for longevity on hash index. Merge HollowPrimaryKeyValueDeriver2.java into HollowUniqueKeyIndex.

Jorel-Master-Of-Scheduling committed:
Take out refactoring of HollowPrimaryKeyIndex since the long term plan is to deprecate it.

Jorel-Master-Of-Scheduling committed:
Fix checkstyle.

dkoszewnik commented on hollow/src/main/java/com/netflix/hollow/core/index/HollowHashIndex.java:
```diff
@@ -216,8 +217,8 @@ public void endUpdate() {
 
-    public HollowReadStateEngine getStateEngine() {
-        return stateEngine;
+    public HollowDataAccess getStateEngine() {
```
I don't think we should change the public API here.

Jorel-Master-Of-Scheduling replied:
Oops! Let me fix that.

dkoszewnik commented on hollow/src/main/java/com/netflix/hollow/core/index/TestableUniqueKeyIndex.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.Collection;
+
+/**
+ * This package is for internal use. Do not depend on it.
```
Should we mark it deprecated?

mbrilnetflix replied:
+1

Jorel-Master-Of-Scheduling replied:
Done.

dkoszewnik reviewed (approved):
Looks good to me, thank you James.

mbrilnetflix commented on hollow/src/main/java/com/netflix/hollow/core/index/key/PrimaryKey.java:
```diff
@@ -31,6 +33,30 @@
+            return new PrimaryKey(type, fieldPaths);
+        }
+
+        HollowSchema schema = hollowDataAccess.getSchema(type);
```
nit: can `hollowDataAccess` be null? it's ok if we have fieldPaths

Jorel-Master-Of-Scheduling replied:
Yes, I think so. Added a comment to the javadoc.

mbrilnetflix reviewed (approved):
LGTM

Jorel-Master-Of-Scheduling committed:
Fix checkstyle.

Jorel-Master-Of-Scheduling committed:
Address PR comments.

Add deprecation to internal interface.

Suppress a few warnings.

Restore API to the original signature.
